bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/evaluation/EvaluationInteractor.kt

```diff
@@ -0,0 +1,108 @@
+  private val latestEvaluationDao: LatestEvaluationDao,
+  private val sharedPrefs: SharedPreferences,
+  private val executor: Executor,
+) {
+  // key: userId
+  @VisibleForTesting
+  internal val latestEvaluations = mutableMapOf<String, List<Evaluation>>()
+
+  @VisibleForTesting
+  internal val currentEvaluations = mutableMapOf<String, List<Evaluation>>()
+
+  @VisibleForTesting
+  internal var currentEvaluationsId: String
+    get() = sharedPrefs.getString(Constants.PREFERENCE_KEY_USER_EVALUATION_ID, "") ?: ""
+    @SuppressLint("ApplySharedPref")
+    set(value) {
+      sharedPrefs.edit()
+        .putString(Constants.PREFERENCE_KEY_USER_EVALUATION_ID, value)
+        .commit()
+    }
+
+  @Suppress("MoveVariableDeclarationIntoWhen")
+  fun fetch(user: User, timeoutMillis: Long?): GetEvaluationsResult {
+    val currentEvaluationsId = this.currentEvaluationsId
+
+    val result = apiClient.getEvaluations(user, currentEvaluationsId, timeoutMillis)
+
+    when (result) {
+      is GetEvaluationsResult.Success -> {
+        val response = result.value.data
+        val newEvaluationsId = response.user_evaluations_id
+        if (currentEvaluationsId == newEvaluationsId) {
+          logd { "Nothing to sync" }
+          return result
+        }
+
+        val newEvaluations = response.evaluations.evaluations
+
+        when (response.state) {
+          UserEvaluationsState.FULL -> {
+            val success = latestEvaluationDao.deleteAllAndInsert(user.id, newEvaluations)
+            if (!success) {
+              loge { "Failed to update latest evaluations" }
+              return result
+            }
+
+            this.currentEvaluationsId = newEvaluationsId
+
+            val featureIds = newEvaluations.map { it.feature_id }
+            currentEvaluationDao.deleteNotIn(user.id, featureIds)
+            val newCurrentEvaluations = currentEvaluationDao.getEvaluations(user.id)
+
+            latestEvaluations[user.id] = newEvaluations
+            currentEvaluations[user.id] = newCurrentEvaluations
+          }
+          UserEvaluationsState.PARTIAL -> {
+            latestEvaluationDao.put(user.id, newEvaluations)
+            latestEvaluations[user.id] = latestEvaluationDao.get(user.id)
+          }
+          UserEvaluationsState.QUEUED -> {
+            // no-op
+          }
+        }
+      }
+      is GetEvaluationsResult.Failure -> {
+        logd(result.error) { "ApiError: ${result.error.message}" }
+      }
+    }
+    return result
+  }
+
+  fun getLatest(userId: String, featureId: String): Evaluation? {
+    val evaluations = latestEvaluations[userId] ?: emptyList()
+    return evaluations.firstOrNull { it.feature_id == featureId }
+  }
+
+  fun getLatestAndRefreshCurrent(userId: String, featureId: String): Evaluation? {
+    val evaluation = getLatest(userId, featureId) ?: return null
+
+    executor.execute {
+      currentEvaluationDao.upsertEvaluation(evaluation)
```

Comment:
The data in the `current_evaluation` table contains only the evaluations that were requested by the app via `BKTClient#boolVariation`, etc. It is used when the `goalEvent` is created.

And the `latest_evaluation` table contains all the latest evaluations got from the `getEvaluations` API.
The interface names are a bit confusing.

The current evaluations are set in the `goalEvent`, but this field is deprecated.
https://github.com/ca-dp/bucketeer-android-sdk/blob/master/bucketeer/src/main/java/jp/bucketeer/sdk/events/EventCreator.kt#L28

I'm looking at the server changes again to ensure that we won't need the current evaluations. If so, the `current_evaluation` table can also be removed.

Reply:
Ah, I blindly followed v1 implementation :(
According to the swagger, indeed GoalEvent request seems not to have evaluations param.

I'm gonna remove current_evaluation table.

Thanks!

Reply:
created a PR #64